### ISSUE TITLE
Bug: attribute starting with `:` formatted differently from the same attribute without `:`

### ISSUE THREAD

**Issue by seletskiy** [prettier/plugin-pug#80]
Hey @Shinigami92, hope you're doing well.

## Info

| Tool     | Version       |
| -------- | ------------- |
| Plugin   | v1.2.0        |
| Prettier | v1.19.1       |
| Node     | v13.12.0       |
| OS       | linux |

## Input

```pug
span(
  :data=`{
    jobs: jobs.all,
    selected: jobs.selected,
    pipelines: [pipeline]
  }`
)
```

## Output or Error

```bash
span(
  :data='{ jobs: jobs.all, selected: jobs.selected, pipelines: [pipeline] }'
)
```

## Expected Output

```pug
span(
  :data=`{
    jobs: jobs.all,
    selected: jobs.selected,
    pipelines: [pipeline]
  }`
)
```

## Additional Context

If `:data` is renamed to `data` (e.g. without `:`), then reformatting doesn't happen.

**Comment by Shinigami92** [comment]
That's because I support Vue out of the box
Long object definitions should be moved to the responsible component

**Comment by Shinigami92** [comment]
If you can live with it, please close this issue
Otherwise, let me know what to do

However, wrapping an object in an attribute is very difficult to implement. That's why I haven't implemented it yet

**Comment by seletskiy** [comment]
@Shinigami92: I'm not quite sure I understand what do you mean.

I use Vue too and instead of `span` there is an actual component, I've just renamed it to `span` in the issue for the simplicity.

From the user standpoint it's a bit mystery why the formatting is different when you include `:` in the attribute name.

Aren't the following two variants works in the same way in Vue (e.g. data object is passed to `child-component`)?

```
child-component(
  :data='{ a: b, c: d }'
)

---

child-component(
  :data=`{
    a: b,
    c: d
  }`
)
```

**Comment by Shinigami92** [comment]
The `tag` is irrelevant, it's only the `tag`'s `attribute` data-binding
So `v-bind:xy` and `:xy` will be formatted with prettier's vue-expression formatting (same as in html)
`xy` (without data-binding) is not a vue-expression and therefore will not be formatted

I'm calling prettier's vue-expression formatting here:  
https://github.com/prettier/plugin-pug/blob/c12fc5933827357f9b3ed2fe7ca5f4b670a7c6db/src/printer.ts#L301
and this is calling this

https://github.com/prettier/plugin-pug/blob/c12fc5933827357f9b3ed2fe7ca5f4b670a7c6db/src/printer.ts#L198
with `parser` being `'__vue_expression'`

**Comment by seletskiy** [comment]
I think there might be some inconsistency.

For example, I have custom Vue directive `v-bem`, which also supports dict as an argument, which I use like this:

```
        div(
          v-bem:+=`{
              pending:  pipeline.status == PipelineStatus.Pending,
              running:  pipeline.status == PipelineStatus.Running,
              failed:   pipeline.status == PipelineStatus.Failed,
              success:  pipeline.status == PipelineStatus.Success,
              canceled: pipeline.status == PipelineStatus.Canceled,
          }`
        )
```

It doesn't get reformatted (and I don't think it should be reformatted) into single-quoted value despite being Vue specific.

I do not want to have automatic object wrapping done by formatter and I understand that it's difficult to implement.

Rather, I think that if attribute is specified in the format `` :blah=`{…}` `` it should be kept as is (e.g. no formatting applied to the its value). That's how it works for my custom `v-bem:...` directive.

What do you think?

**Comment by Shinigami92** [comment]
Using \` after `=` in pug is different from using `'` or `"`! When using \`, it will be consumed by pug and parsed as JavaScript.
You should move your object into a computed value

```pug
div(v-bem:+="state")
```

```js
/*...*/ {
  /*...*/
  computed: {
    state() {
      return {
        pending:  pipeline.status == PipelineStatus.Pending,
        running:  pipeline.status == PipelineStatus.Running,
        failed:   pipeline.status == PipelineStatus.Failed,
        success:  pipeline.status == PipelineStatus.Success,
        canceled: pipeline.status == PipelineStatus.Canceled
      }
    }
  }
}
```

Or to make it much shorter

```js
return { [pipeline.status]: true } // other values are undefined / falsy
// may be transformed to lowercase
```

With the computed value you also benefit from rendercycles, cause its cached

**Comment by seletskiy** [comment]
@Shinigami92: thanks for the tips, I appreciate.

> When using `, it will be consumed by pug and parsed as JavaScript.

Are you sure about it? [Pug documentation][1] doesn't say that. It says that using `` ` `` is useful for putting multiline values in the attributes and that's very convenient way for using [object litlerals][2] as described in Vue docs.

> Or to make it much shorter

I've put an inappropriate example to illustrate the issue and it can be shortened using your tip indeed, but there are other examples where multiline object literals are useful, like this one:

```pug
tr(
  v-bem:item,
  v-bem:+=`{
    navigable:  true,
    selectable: !!data,
    selected:   selected && selected.id == job.id,
  }`,
  v-for='job in jobs'
)
```

> You should move your object into a computed value

Computed property is not appropriate in this case, since this particular directive is used to describe suffixes in CSS class names (modifiers in BEM model) for some particular HTML element and it's not inherit property of the component itself, so named attribute will be artificial.

Anyway, it feels that discussion is not moving in the correct direction.

To recap my point:

1. `` ` ``-quoted values are used to write multiline attribute values in Pug **AND**
2. Object literals are useful to inline state in the Vue templates **THEN**
3. Source code formatter should allow to do that `v-bind` directive **GIVEN THAT**.
4. Source code formatter already allow to do that for custom directives.

[1]: https://pugjs.org/language/attributes.html#multiline-attributes
[2]: https://vuejs.org/v2/guide/custom-directive.html#Object-Literals

**Comment by Shinigami92** [comment]
> Are you sure about it? [Pug documentation](https://pugjs.org/language/attributes.html#multiline-attributes) doesn't say that. It says that using `` ` `` is useful for putting multiline values in the attributes and that's very convenient way for using [object litlerals](https://vuejs.org/v2/guide/custom-directive.html#Object-Literals) as described in Vue docs.

I'm not so sure right now anymore 🤔
but I thought I remember having big problems with it in a different context in the past 🤷‍♂️

---

> I've put an inappropriate example to illustrate the issue and it can be shortened using your tip indeed, but there are other examples where multiline object literals are useful, like this one:
>
> ```
> tr(
>   v-bem:item,
>   v-bem:+=`{
>     navigable:  true,
>     selectable: !!data,
>     selected:   selected && selected.id == job.id,
>   }`,
>   v-for='job in jobs'
> )
> ```
>
> Computed property is not appropriate in this case, since this particular directive is used to describe suffixes in CSS class names (modifiers in BEM model) for some particular HTML element and it's not inherit property of the component itself, so named attribute will be artificial.

You are right, you shouldn't use a computed value for the new example.
But why not use a method?

```pug
tr(v-for='job in jobs', v-bem:item, v-bem:+="rowBem(job)")
```

```js
/* ... */ {
  /* ... */,
  methods: {
    rowBem(job) {
      return {
        navigable:  true,
        selectable: !!data,
        selected:   selected?.id == job.id,
      }
    }
  }
}
```

In this way, you extract the calculation logic from your view and also benefit from better readability.
On top of this you could benefit from TypeScript or JSDoc typing and introduce type safety ✨

---

> Anyway, it feels that discussion is not moving in the correct direction.
>
> To recap my point:
>
> 1. `` ` ``-quoted values are used to write multiline attribute values in Pug **AND**
> 2. Object literals are useful to inline state in the Vue templates **THEN**
> 3. Source code formatter should allow to do that `v-bind` directive **GIVEN THAT**.
> 4. Source code formatter already allow to do that for custom directives.

I know, I know...
Sorry for all of these workarounds and stuff. I know it can be a bit frustrating at first, but in my opinion, not only are there already workarounds, it also makes the developer experience better to use these workarounds.
It would also be very difficult to support this and not introduce bugs for e.g. line-length checks (and this is already complicated).

To point 4. I should say that I havent found all directives like `v-bem` yet.
Regarding point 4, I should say that I have not yet found all the vue directives, such as `v-bem`.
I check it currently in this line:

https://github.com/prettier/plugin-pug/blob/c12fc5933827357f9b3ed2fe7ca5f4b670a7c6db/src/utils/vue.ts#L51

So it may be that I should add `v-bem` also to this check.

To shorten this a little: I would always be happy to check and merge good pull requests!
I am also ready to help you implement the code and help you find the right lines and problems in the code.
I would also leave a message in README.md that you, as a contributor, have helped to support this.

**Comment by Shinigami92** [comment]
I have released a [v1.4.2](https://www.npmjs.com/package/@prettier/plugin-pug/v/1.4.2) 🎉
